### PR TITLE
Implement clearing application related caches when adding, updating the name or deleting a role.

### PR DESCRIPTION
**Pull request by nipunsampath**
### Proposed changes in this pull request
This PR 

- Resolves https://github.com/wso2/product-is/issues/18509 by implementing cache clear for `ServiceProviderByIDCache` after adding a role.
- Implement clearing application related caches when updating the name and deleting a role.

### Related Issues

- https://github.com/wso2/product-is/issues/18509

### Follow up actions

- The integration test `OrganizationManagementSuccessTest` should be re-enabled. Ref: https://github.com/wso2/product-is/pull/18504

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java [hunk 1]
```diff
@@ -180,6 +182,18 @@ public void preUpdateRoleName(String roleID, String newRoleName, String tenantDo
...
+        try {
+            Role role = ApplicationManagementServiceComponentHolder.getInstance().getRoleManagementServiceV2()
+                    .getRole(roleID, tenantDomain);
+            List<AssociatedApplication> associatedApplications = role.getAssociatedApplications();
+            for (AssociatedApplication application : associatedApplications) {
+                clearApplicationCaches(application.getId(), tenantDomain);
+            }
+        } catch (IdentityRoleManagementException e) {
+            throw new IdentityRoleManagementException(
+                    String.format("Error occurred while updating the name of role : %s and tenant domain : %s",
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java:194]:
```suggestion
                    String.format("Error occurred while updating the name of role : %s in tenant domain : %s",
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java [hunk 2]
```diff
@@ -525,6 +537,56 @@ private String getApplicationName(String applicationID, String tenantDomain)
...
+            throws IdentityRoleManagementServerException {
+
+        try {
+            return ApplicationManagementService.getInstance()
+                    .getApplicationBasicInfoByResourceId(applicationID, tenantDomain);
+        } catch (IdentityApplicationManagementException e) {
+            String errorMessage = "Error while retrieving the basic information for the given app id: " + applicationID;
+            throw new IdentityRoleManagementServerException(UNEXPECTED_SERVER_ERROR.getCode(), errorMessage, e);
+        }
+
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java:559]:
```suggestion
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java [hunk 3]
```diff
@@ -525,6 +537,56 @@ private String getApplicationName(String applicationID, String tenantDomain)
...
+        ServiceProviderResourceIdCacheKey resourceIdKey = new ServiceProviderResourceIdCacheKey(appId);
+        ServiceProviderByResourceIdCache.getInstance().clearCacheEntry(resourceIdKey, tenantDomain);
+        
+        ServiceProviderIDCacheKey appIdKey = new ServiceProviderIDCacheKey(appBasicInfo.getApplicationId());
+        ServiceProviderByIDCache.getInstance().clearCacheEntry(appIdKey, tenantDomain);
+        
+        IdentityServiceProviderCacheKey appNameKey = new IdentityServiceProviderCacheKey(
+                appBasicInfo.getApplicationName());
+        IdentityServiceProviderCache.getInstance().clearCacheEntry(appNameKey, tenantDomain);
+
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java:587]:
```suggestion
```

**Review by AnuradhaSK — DISMISSED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java (hunk 3 above)
Comment (nipunsampath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java]: Fixed with cbd98834db6aea52968651f23983562fe14c0d87

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java (hunk 2 above)
Comment (nipunsampath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java]: Fixed with cbd98834db6aea52968651f23983562fe14c0d87

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java (hunk 1 above)
Comment (nipunsampath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java]: Fixed with cbd98834db6aea52968651f23983562fe14c0d87

**Review by AnuradhaSK — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java [hunk 4]
```diff
@@ -180,6 +182,18 @@ public void preUpdateRoleName(String roleID, String newRoleName, String tenantDo
     public void postUpdateRoleName(String roleID, String newRoleName, String tenantDomain)
             throws IdentityRoleManagementException {
 
+        try {
+            Role role = ApplicationManagementServiceComponentHolder.getInstance().getRoleManagementServiceV2()
+                    .getRole(roleID, tenantDomain);
+            List<AssociatedApplication> associatedApplications = role.getAssociatedApplications();
+            for (AssociatedApplication application : associatedApplications) {
+                clearApplicationCaches(application.getId(), tenantDomain);
```
[the remark this answers is not in the document]
Comment (nipunsampath) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/DefaultRoleManagementListener.java:184]: Fixed with d38c68f5674f4b3649a1bb3857e8ee85d5b59309.

**Review by dewniMW — APPROVED** [review]

**Pull request merged**